functions/index.js

```diff
@@ -1,17 +1,139 @@
 const { onSchedule } = require("firebase-functions/v2/scheduler");
 const { setGlobalOptions } = require("firebase-functions/v2");
 const { initializeApp } = require("firebase-admin/app");
+initializeApp();
 const { getFirestore } = require("firebase-admin/firestore");
 const logger = require("firebase-functions/logger");
+const { onDocumentUpdated } = require("firebase-functions/v2/firestore");
+const { notifyStudentLessonConfirmedByTutor, notifyTutorLessonConfirmedByStudent, notifyStudentLessonPendingReview, notifyStudentAsTutorOfferedToTeach} = require("./lessonNotifications");
 
-initializeApp();
 
 setGlobalOptions({
     maxInstances: 10,
     timeoutSeconds: 60,
     memory: "256MiB"
 });
 
+//Send a notification to the Tutor or Student when a lesson is updated
+exports.LessonUpdateNotification = onDocumentUpdated("lessons/{lessonId}", async (event) => {
+  const before = event.data.before.data();
+  const after = event.data.after.data();
+
+  if (before.status !== after.status) {
+    console.log(`Lesson status changed from ${before.status} to ${after.status}`);
+
+    // Notify student when lesson is confirmed by tutor
+    if (before.status === "PENDING_TUTOR_CONFIRMATION" && after.status === "CONFIRMED") {
+      await notifyStudentLessonConfirmedByTutor(after);
+    }
+
+    // Notify tutor when lesson is confirmed by student
+    if (before.status === "STUDENT_REQUESTED" && after.status === "CONFIRMED") {
+      await notifyTutorLessonConfirmedByStudent(after);
+    }
+
+    // Notify student when lesson is pending review
+    if (before.status === "CONFIRMED" && after.status === "PENDING_REVIEW") {
+      await notifyStudentLessonPendingReview(after);
+    }
+  }
+
+  //Detect when an additional tutor offer to teach a requested lesson:
+  //more specifically when a tutor uid is added to the tutorUid list of a STUDENT_REQUESTED lesson
+  if (after.status === "STUDENT_REQUESTED" && before.tutorUid.length < after.tutorUid.length) {
+    console.log("Tutor added to the lesson's tutorUid.");
+    await notifyStudentAsTutorOfferedToTeach(after);
+  }
+});
+
+
+exports.lessonReminder = onSchedule("every 5 minutes", async (event) => {
+  const now = new Date();
+  const oneHourFromNow = new Date(now.getTime() + 60 * 60 * 1000);
```

Comment:
The code follows a clear flow formatting dates, querying Firestore, processing results, and sending notifications.